lib/recursive-methods.rb

```diff
@@ -1,49 +1,91 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n), err, does the stack take up a lot of space? theres n numbers of calls
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError, "cant be negative number" if 0 > n
+  if n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n) where n is the length of the string
+# Space complexity: o(n) because it creates "temp" string to return
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s == ""
+  if s.length == 1
+    return s
+  else
+    return temp = s[-1] + reverse(s[0...-1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: constant, cause its in place, although, once again, it makes s.length stack calls, so does that mean its O(n) because it takes a place in the stack that many times?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s == ""
+  return s[-1] + reverse_inplace(s[0...-1])
```

Comment:
This is creating new strings, so it isn't in place. 

To do it in place you need to swap characters one at a time without creating a new string.